### PR TITLE
Docs. Initial version.

### PR DESCRIPTION
**Pull request by s-andrus**
This is the initial version of the extended task description i.e. how I got the task.
Also, there some initial design decisions as well.

### CODE REVIEW

File: docs/design.txt [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+Also optimization like indexing of the logs lines are out of scope.
+
+ Job result code.
+Job result code is stored in the corresponnding file along with 
+the command meta information:
+ - command name, command params
+ - start time, finish time
+
+ Reliability.
+If the scheduler is died some information may be lost after the service restart.
```
Comment (russjones) [inline docs/design.txt]: What would be lost? If I understand correctly you would be able to show all completed jobs but not in-progress jobs?

Is this a correct understanding?

File: docs/design.txt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+A client certificates/keys shall be distributed properly and it is not 
+a part of this task.
+  Clients part.
+So every user must have its own pair of certificates/keys 
+and keep the keys secure.
+  Server part.
+The server has all the client certificates and only clients who provides 
+a certificate from that preloaded certificate list are alowed to connect 
+to the server.
+Alternativelly, any certificate which is signed by a trusted authority
```
Comment (russjones) [inline docs/design.txt]: I think just making sure the certificate presented is signed by a known CA is enough.

File: docs/design.txt [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+for the request deduplication.
+TBD: The proto file will be provided later.
+
+ Clients authentication.
+The regular TLS certificates are used for authentication. 
+A client certificates/keys shall be distributed properly and it is not 
+a part of this task.
+  Clients part.
+So every user must have its own pair of certificates/keys 
+and keep the keys secure.
```
Comment (russjones) [inline docs/design.txt]: Can you talk more about how the client certificates will be generated and used?

File: docs/design.txt (hunk 1 above)
Comment (s-andrus) [reply docs/design.txt]: Yeah, that's correct.

File: docs/design.txt (hunk 2 above)
Comment (s-andrus) [reply docs/design.txt]: The idea is to use the certificates for authentication so the service needs to map a certificate to a user name somehow. As an option, a user name can be added to the certificate and signed by a known CA so  that's enough. And also checking a revocation list is a must as well.

File: docs/design.txt (hunk 3 above)
Comment (s-andrus) [reply docs/design.txt]: The idea is to generate the keys/certificates on my own and just distribute between the clients and service. So the service trusts only to known certificates and nothing else i.e. there is no known CA. For the simplicity all the users are mapped to the same Linux OS user which is the scheduler user. So the only commands which can run scheduler are allowed and there is no per client user in the first version.

Comment (s-andrus) [reply docs/design.txt]: I.e. a client configured with: a client private key (it is a secret), a corresponding client certificate, and a service certificate. The service configured with: a service private key, a corresponding service certificate, and a list of all the client certificates.

Comment (russjones) [reply docs/design.txt]: > i.e. there is no known CA

> and a list of all the client certificates

Not sure I fully understand the above parts.

You are planning on storing multiple client certificates (self signed) on the service and accept connections if it comes from one of the known client certificates?

Comment (s-andrus) [reply docs/design.txt]: Self signed certificates with whitelisting can be an option but may put to much pressure on the distribution side.
The idea is: 
    - to sign the certificates by known authority
    - to whitelist explicitly what certificates the peers trust even if these are properly signed by known authority. I prepared PR2 as a prove of this concept,

**Pull request merged**